Quarkus 3.18.1 can't find config inner classes during @QuarkusTest tests

### Describe the bug

I have a [CommaFeedConfiguration](https://github.com/Athou/commafeed/blob/741fe71dc60559f2d130a669a6e9a09d9cfda043/commafeed-server/src/main/java/com/commafeed/CommaFeedConfiguration.java) configuration class. Quarkus 3.28.1 cannot find its inner classes during tests annotated with `@QuarkusTest`

`java.lang.ClassNotFoundException: com.commafeed.CommaFeedConfiguration$HttpClientCache`

### Expected behavior

Tests pass as they did with Quarkus 3.26.4

### Actual behavior

Here's the stacktrace during the maven build : 

```
[ERROR] Tests run: 4, Failures: 0, Errors: 1, Skipped: 3, Time elapsed: 5.043 s <<< FAILURE! -- in com.commafeed.e2e.AuthentificationIT
[ERROR] com.commafeed.e2e.AuthentificationIT.registerSuccess -- Time elapsed: 0.005 s <<< ERROR!
java.lang.RuntimeException: java.lang.NoClassDefFoundError: com/commafeed/CommaFeedConfiguration$HttpClientCache
        at io.quarkus.test.junit.QuarkusTestExtension.throwBootFailureException(QuarkusTestExtension.java:666)
        at io.quarkus.test.junit.QuarkusTestExtension.interceptTestClassConstructor(QuarkusTestExtension.java:761)
        at java.base/java.util.Optional.orElseGet(Optional.java:364)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1604)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1604)
Caused by: java.lang.NoClassDefFoundError: com/commafeed/CommaFeedConfiguration$HttpClientCache
        at java.base/java.lang.ClassLoader.defineClass1(Native Method)
        at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:962)
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:566)
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:524)
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:577)
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:524)
        at io.quarkus.hibernate.validator.runtime.HibernateBeanValidationConfigValidatorBuilder.<clinit>(Unknown Source)
        at io.quarkus.runtime.generated.StaticInitConfig.configBuilder(Unknown Source)
        at io.quarkus.runtime.configuration.AbstractConfigBuilder.withCustomizer(AbstractConfigBuilder.java:117)
        at io.quarkus.runtime.generated.Config.<clinit>(Unknown Source)
        at io.quarkus.runner.ApplicationImpl.<clinit>(Unknown Source)
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Class.java:547)
        at io.quarkus.runner.bootstrap.StartupActionImpl.run(StartupActionImpl.java:334)
        at io.quarkus.test.junit.QuarkusTestExtension.doJavaStart(QuarkusTestExtension.java:232)
        at io.quarkus.test.junit.QuarkusTestExtension.ensureStarted(QuarkusTestExtension.java:640)
        at io.quarkus.test.junit.QuarkusTestExtension.beforeAll(QuarkusTestExtension.java:685)
        ... 1 more
        Suppressed: java.lang.NoClassDefFoundError: Could not initialize class io.quarkus.runtime.generated.Config
                at java.base/java.lang.Class.forName0(Native Method)
                at java.base/java.lang.Class.forName(Class.java:547)
                at io.quarkus.runner.bootstrap.StartupActionImpl.run(StartupActionImpl.java:338)
                ... 4 more
        Caused by: java.lang.ExceptionInInitializerError: Exception java.lang.NoClassDefFoundError: com/commafeed/CommaFeedConfiguration$HttpClientCache [in thread "main"]
                at java.base/java.lang.ClassLoader.defineClass1(Native Method)
                at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:962)
                at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:566)
                at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:524)
                at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:577)
                at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:524)
                at io.quarkus.hibernate.validator.runtime.HibernateBeanValidationConfigValidatorBuilder.<clinit>(Unknown Source)
                at io.quarkus.runtime.generated.StaticInitConfig.configBuilder(Unknown Source)
                at io.quarkus.runtime.configuration.AbstractConfigBuilder.withCustomizer(AbstractConfigBuilder.java:117)
                at io.quarkus.runtime.generated.Config.<clinit>(Unknown Source)
                at io.quarkus.runner.ApplicationImpl.<clinit>(Unknown Source)
                at java.base/java.lang.Class.forName0(Native Method)
                at java.base/java.lang.Class.forName(Class.java:547)
                at io.quarkus.runner.bootstrap.StartupActionImpl.run(StartupActionImpl.java:334)
                ... 4 more
Caused by: java.lang.ClassNotFoundException: com.commafeed.CommaFeedConfiguration$HttpClientCache
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:550)
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:524)
        ... 18 more
```

### How to Reproduce?

Checkout [this branch](https://github.com/Athou/commafeed/pull/1904) and run `mvn clean install`

### Output of `uname -a` or `ver`

Microsoft Windows [Version 10.0.26100.6584]

### Output of `java -version`

Java version: 17.0.11, vendor: Oracle Corporation

### Quarkus version or git rev

3.28.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.11

### Additional information

_No response_

--- a/core/deployment/src/main/java/io/quarkus/deployment/configuration/ConfigMappingUtils.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/configuration/ConfigMappingUtils.java
@@ -1,5 +1,6 @@
 package io.quarkus.deployment.configuration;
 
+import static io.quarkus.bootstrap.classloading.QuarkusClassLoader.isApplicationClass;
 import static io.smallrye.config.ConfigMappings.ConfigClass.configClass;
 import static org.jboss.jandex.AnnotationTarget.Kind.CLASS;
 
@@ -21,12 +22,12 @@ import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
 import io.quarkus.deployment.builditem.ConfigClassBuildItem;
 import io.quarkus.deployment.builditem.ConfigClassBuildItem.Kind;
+import io.quarkus.deployment.builditem.ConfigurationBuildItem;
 import io.quarkus.deployment.builditem.GeneratedClassBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveClassBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveMethodBuildItem;
 import io.quarkus.deployment.util.ReflectUtil;
 import io.quarkus.hibernate.validator.spi.AdditionalConstrainedClassBuildItem;
-import io.quarkus.runtime.annotations.ConfigRoot;
 import io.smallrye.config.ConfigMapping;
 import io.smallrye.config.ConfigMappingInterface;
 import io.smallrye.config.ConfigMappingInterface.LeafProperty;
@@ -38,7 +39,6 @@ import io.smallrye.config.ConfigMappings.ConfigClass;
 
 public class ConfigMappingUtils {
     public static final DotName CONFIG_MAPPING_NAME = DotName.createSimple(ConfigMapping.class.getName());
-    public static final DotName CONFIG_ROOT_NAME = DotName.createSimple(ConfigRoot.class.getName());
 
     private ConfigMappingUtils() {
         throw new UnsupportedOperationException();
@@ -46,6 +46,7 @@ public class ConfigMappingUtils {
 
     // Used for application Mappings and MP ConfigProperties
     public static void processConfigClasses(
+            ConfigurationBuildItem configItem,
             CombinedIndexBuildItem combinedIndex,
             BuildProducer<GeneratedClassBuildItem> generatedClasses,
             BuildProducer<ReflectiveClassBuildItem> reflectiveClasses,
@@ -62,27 +63,31 @@ public class ConfigMappingUtils {
                 continue;
             }
 
-            // Skip roots
-            if (target.hasAnnotation(CONFIG_ROOT_NAME)) {
+            // Ignore scanned classes from the processor.
+            // Usually they are not available in Jandex, but there are cases where extensions have jandex to their
+            // own classes, or user applications adding @ConfigRoot to generate documentation. In that case, it would
+            // generate duplicates.
+            ConfigClass configClass = configClass(
+                    toClass(target.asClass().name()),
+                    Optional.ofNullable(annotationPrefix).map(AnnotationValue::asString).orElse(""));
+            if (configItem.getReadResult().getAllMappingsByClass().containsKey(configClass.getType())) {
                 continue;
             }
-
-            Class<?> configClass = toClass(target.asClass().name());
-            String prefix = Optional.ofNullable(annotationPrefix).map(AnnotationValue::asString).orElse("");
             Kind configClassKind = getConfigClassType(instance);
-            processConfigClass(configClass(configClass, prefix), configClassKind, true, combinedIndex,
-                    generatedClasses, reflectiveClasses, reflectiveMethods, configClasses, additionalConstrainedClasses);
+            processConfigClass(configClass, configClassKind, combinedIndex, generatedClasses, reflectiveClasses,
+                    reflectiveMethods, configClasses, additionalConstrainedClasses);
         }
     }
 
     public static void processConfigMapping(
+            ConfigurationBuildItem configItem,
             CombinedIndexBuildItem combinedIndex,
             BuildProducer<GeneratedClassBuildItem> generatedClasses,
             BuildProducer<ReflectiveClassBuildItem> reflectiveClasses,
             BuildProducer<ReflectiveMethodBuildItem> reflectiveMethods,
             BuildProducer<ConfigClassBuildItem> configClasses,
             BuildProducer<AdditionalConstrainedClassBuildItem> additionalConstrainedClasses) {
-        processConfigClasses(combinedIndex, generatedClasses, reflectiveClasses, reflectiveMethods, configClasses,
+        processConfigClasses(configItem, combinedIndex, generatedClasses, reflectiveClasses, reflectiveMethods, configClasses,
                 additionalConstrainedClasses, CONFIG_MAPPING_NAME);
     }
 
@@ -95,14 +100,13 @@ public class ConfigMappingUtils {
             BuildProducer<ConfigClassBuildItem> configClasses,
             BuildProducer<AdditionalConstrainedClassBuildItem> additionalConstrainedClasses) {
 
-        processConfigClass(configClass, Kind.MAPPING, false, combinedIndex, generatedClasses, reflectiveClasses,
+        processConfigClass(configClass, Kind.MAPPING, combinedIndex, generatedClasses, reflectiveClasses,
                 reflectiveMethods, configClasses, additionalConstrainedClasses);
     }
 
     private static void processConfigClass(
             ConfigClass configClassWithPrefix,
             Kind configClassKind,
-            boolean isApplicationClass,
             CombinedIndexBuildItem combinedIndex,
             BuildProducer<GeneratedClassBuildItem> generatedClasses,
             BuildProducer<ReflectiveClassBuildItem> reflectiveClasses,
@@ -121,7 +125,8 @@ public class ConfigMappingUtils {
             generatedClassesNames.add(mappingMetadata.getClassName());
             // This is the generated implementation of the mapping by SmallRye Config.
             byte[] classBytes = mappingMetadata.getClassBytes();
-            generatedClasses.produce(new GeneratedClassBuildItem(isApplicationClass, mappingMetadata.getClassName(),
+            generatedClasses.produce(new GeneratedClassBuildItem(isApplicationClass(configClass.getName()),
+                    mappingMetadata.getClassName(),
                     classBytes));
             additionalConstrainedClasses.produce(AdditionalConstrainedClassBuildItem.of(mappingMetadata.getClassName(),
                     classBytes));

--- a/core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigGenerationBuildStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigGenerationBuildStep.java
@@ -204,7 +204,7 @@ public class ConfigGenerationBuildStep {
             BuildProducer<ConfigClassBuildItem> configClasses,
             BuildProducer<AdditionalConstrainedClassBuildItem> additionalConstrainedClasses) {
 
-        processConfigMapping(combinedIndex, generatedClasses, reflectiveClasses, reflectiveMethods, configClasses,
+        processConfigMapping(configItem, combinedIndex, generatedClasses, reflectiveClasses, reflectiveMethods, configClasses,
                 additionalConstrainedClasses);
 
         List<ConfigClass> buildTimeRunTimeMappings = configItem.getReadResult().getBuildTimeRunTimeMappings();

--- a/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ConfigBuildStep.java
+++ b/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ConfigBuildStep.java
@@ -63,6 +63,7 @@ import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
 import io.quarkus.deployment.builditem.ConfigClassBuildItem;
 import io.quarkus.deployment.builditem.ConfigMappingBuildItem;
 import io.quarkus.deployment.builditem.ConfigPropertiesBuildItem;
+import io.quarkus.deployment.builditem.ConfigurationBuildItem;
 import io.quarkus.deployment.builditem.GeneratedClassBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveClassBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveMethodBuildItem;
@@ -261,6 +262,7 @@ public class ConfigBuildStep {
 
     @BuildStep
     void generateConfigProperties(
+            ConfigurationBuildItem configItem,
             CombinedIndexBuildItem combinedIndex,
             BuildProducer<GeneratedClassBuildItem> generatedClasses,
             BuildProducer<ReflectiveClassBuildItem> reflectiveClasses,
@@ -268,7 +270,7 @@ public class ConfigBuildStep {
             BuildProducer<ConfigClassBuildItem> configClasses,
             BuildProducer<AdditionalConstrainedClassBuildItem> additionalConstrainedClasses) {
 
-        processConfigClasses(combinedIndex, generatedClasses, reflectiveClasses, reflectiveMethods, configClasses,
+        processConfigClasses(configItem, combinedIndex, generatedClasses, reflectiveClasses, reflectiveMethods, configClasses,
                 additionalConstrainedClasses, MP_CONFIG_PROPERTIES_NAME);
     }
 

